File: leetcode/lc-us/cpp/1145.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int left, right;

    int dfs(TreeNode *root, int x) {
        if (!root) return 0;
        auto l = dfs(root->left, x), r = dfs(root->right, x);
        if (root->val == x) {
            left = l, right = r;
        }
        return l + r + 1;
    }

public:
    bool btreeGameWinningMove(TreeNode *root, int n, int x) {
        left = 0, right = 0;
        dfs(root, x);
        int up = n - left - right - 1;
        if (up > left + right || left > right + up || right > left + up) {
            return true;
        }
        return false;
    }
};

File: leetcode/lc-us/py/1145.py
from collections import defaultdict

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def btreeGameWinningMove(self, root: Optional[TreeNode], n: int, x: int) -> bool:
        g = defaultdict(list)

        def dfs1(node: Optional[TreeNode]) -> None:
            if not node:
                return
            if node.left:
                g[node.val].append(node.left.val)
                g[node.left.val].append(node.val)
                dfs1(node.left)
            if node.right:
                g[node.val].append(node.right.val)
                g[node.right.val].append(node.val)
                dfs1(node.right)

        dfs1(root)

        def dfs2(u: int, p: int) -> int:
            res = 0
            for ne in g[u]:
                if ne != p:
                    res += dfs2(ne, u)
            return res + 1

        for y in g[x]:
            t = dfs2(y, x)
            if t > n // 2:
                return True
        return False
